creusot-contracts/src/ghost_ptr.rs

```diff
@@ -140,6 +146,100 @@ impl<T: ?Sized> GhostPtrExt<T> for GhostPtr<T> {
     }
 }
 
+impl<'a, T: ?Sized> ShallowModel for GhostPtrTokenRef<'a, T> {
+    type ShallowModelTy = FMap<GhostPtr<T>, T>;
+
+    #[trusted]
+    #[ghost]
+    #[open(self)]
+    fn shallow_model(self) -> Self::ShallowModelTy {
+        absurd
+    }
+}
+
+impl<'a, T: ?Sized> Deref for GhostPtrTokenRef<'a, T> {
+    type Target = GhostPtrToken<T>;
+
+    #[trusted]
+    #[ensures(result@ == self@)]
+    fn deref(&self) -> &Self::Target {
+        &GhostPtrToken(PhantomData)
+    }
+}
+
+impl<'a, T: ?Sized> GhostPtrTokenRef<'a, T> {
+    /// Shrinks the view of the `self` so that it's model is now new-model
+    #[trusted]
+    #[requires(_new_model.subset(self@))]
+    #[ensures(result@ == *_new_model)]
+    pub fn shrink_token_ref(self, _new_model: Ghost<FMap<*const T, T>>) -> Self {
+        self
+    }
+}
+
+impl<'a, T: ?Sized> GhostPtrTokenMut<'a, T> {
+    #[trusted]
+    #[ghost]
+    #[open(self)]
+    pub fn curr(self) -> FMap<GhostPtr<T>, T> {
+        absurd
+    }
+
+    #[trusted]
+    #[logic]
+    #[open(self)]
+    pub fn fin(self) -> FMap<GhostPtr<T>, T> {
+        absurd
+    }
+
+    #[ensures(self.fin() == self.curr())]
+    #[ensures(result@ == self.curr())]
+    pub fn shr(self) -> GhostPtrTokenRef<'a, T> {
+        GhostPtrTokenRef(PhantomData)
+    }
+
+    /// Mutably borrows `ptr` and shrinks `t` so that it can no longer be used to access `ptr`
```

Comment:
I `t` should be `self` here?

Reply:
Also, given that this is the most complex function in the API by far, do you think you could include a simple example usage? Each time I read this it takes me a few minutes to understand how it works again.